Sources/MongoSwift/MongoError.swift

```diff
@@ -358,34 +387,42 @@ internal func extractBulkWriteError<T: Codable>(
 }
 
 /// Extracts a `WriteConcernError` from a server reply.
-private func extractWriteConcernError(from reply: Document) throws -> WriteConcernError? {
+private func extractWriteConcernError(from reply: Document) throws -> WriteConcernFailure? {
     guard let writeConcernErrors = reply["writeConcernErrors"]?.arrayValue?.compactMap({ $0.documentValue }),
         !writeConcernErrors.isEmpty else {
         return nil
     }
-    return try BSONDecoder().decode(WriteConcernError.self, from: writeConcernErrors[0])
+    return try BSONDecoder().decode(WriteConcernFailure.self, from: writeConcernErrors[0])
 }
 
 /// Internal function used by write methods performing single writes that are implemented via the bulk API. Catches any
-/// ServerError.bulkWriteErrors thrown by the given closure and converts them to ServerError.writeErrors. All other
+/// BulkWriteErrors thrown by the given closure and converts them to WriteErrors. All other
 /// errors will be propagated as-is.
 internal func convertingBulkWriteErrors<T>(_ body: () throws -> T) throws -> T {
     do {
         return try body()
-    } catch let ServerError.bulkWriteError(bulkWriteErrors, writeConcernError, other, _, errorLabels) {
-        if let bwes = bulkWriteErrors, !bwes.isEmpty {
-            let writeError = WriteError(
-                code: bwes[0].code,
-                codeName: bwes[0].codeName,
-                message: bwes[0].message
+    } catch let bwe as BulkWriteError {
+        let writeFailure: WriteFailure? = bwe.writeFailures.flatMap { failures in
```

Comment:
I think I am a little confused about the use of flatmap here.... doesn't flatmap return an array?

Reply:
ahhh, I see, tricky. thanks for the explanation.